### PR TITLE
ex3 initial commit

### PR DESCRIPTION
**Pull request by Ouriel91**

### CODE REVIEW

**Comment by Ouriel91** [comment]
@SergeiSafrigin

**Comment by Ouriel91** [comment]
@SergeiSafrigin @shaneer

File: src/ex3/README.md [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Exercise 3
+For run code on local machine:
+just run the command *node index* in the ex3 folder in terminal, after then will open visual cli
+
+for install my package type in your folder project terminal
+1. (npm init first of all)
+2. npm ouriel-todolist-cli
+3. npx ouriel-todolist-cli
```
Comment (shaneer) [inline src/ex3/README.md:8]: I like that you updated the readme to be specific to you :)

File: src/ex3/todolist-model.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
+import fs from 'fs';
```
Comment (shaneer) [inline src/ex3/todolist-model.js]: you need to separate your code into more specific files/services

For example you can have a data access service or layer, in this case a service that has functions that read/write to the file and separate this from here.

Having really long fliles makes the code less readable but more importantly less maintainable and leads to more bugs when multiple people work on the same code

File: src/ex3/todolist-model.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+            if(data === null){
+                this.todoList = []
+            }
+            else{
+                this.todoList = JSON.parse(data)
+            } 
+        } catch (err) {
+            console.error(err);
+        }
+    }
```
Comment (shaneer) [inline src/ex3/todolist-model.js:19]: You should use some kind of linting tool, you can use the prettier extension in vscode for example.
This will autocorrect on save so you don't have spacing inconsistencies or some statements with ; and some without.

File: src/ex3/todolist-model.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+    addData(enterValue){
+        this.todoList.push(enterValue)
+    }
+
+    removeData(index){
+        let copyArr = this.todoList
+        copyArr.splice(index,1)
+        this.todoList = copyArr
+    }
```
Comment (shaneer) [inline src/ex3/todolist-model.js:29]: these could be seperted out into a utils module which would make sense and make this file cleaner

File: src/ex3/todolist-model.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+
+    orderDataAlphabeticallyReverse(){
+        this.todoList = TodoListModel.compareFunc(this.todoList,"title").reverse()
+    }
+
+    orderUnDoneToDone(){
+        this.todoList = TodoListModel.compareFunc(this.todoList,"done")
+    }
+
+    orderDoneToUnDone(){
```
Comment (shaneer) [inline src/ex3/todolist-model.js:51]: I like the reusablility fo the compareFunc here

File: src/ex3/cli-methods.js [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+  
+    data.forEach((todo, index) => {
+        const indexText = chalk.bold.cyanBright(index)
+        const titleText = chalk.bold.magenta(todo.title)
+        const doneUndoneText = todo.done ? chalk.bold.green("done") : chalk.bold.red("undone")
+        console.log(`${indexText}: ${titleText} is ${doneUndoneText}`)
+    })
+}
+
+export function numberValidation(value){
```
Comment (shaneer) [inline src/ex3/cli-methods.js]: a better name would be isValidNumber then it makes sense that it returns a bool

File: src/ex3/cli-methods.js [hunk 7]
```diff
@@ -0,0 +1,302 @@
...
+    })
+}
+
+export function deleteTodoFromList() {
+  inquirer.prompt({
+    type: 'number',
+    name: 'deleteTodo',
+    message: 'enter the index for todo that you want to delete', 
+  })
+  .then(answers => {
```
Comment (shaneer) [inline src/ex3/cli-methods.js:95]: would be better to seperatet his into a named function , preferably in a different file, this way it is a named function and not an anonymous function, but also amkes the code a lot more readable

File: src/ex3/cli-methods.js [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+        todos = main.getUnDoneTodos()
+        break;
+    }
+
+    getFullCurrentList(todos)
+  })
+}
+
+export function displayPokemon() {
+
```
Comment (shaneer) [inline src/ex3/cli-methods.js:283]: the UX here is rather harsh since the only way to display index is by done/undone and then it's hard to know whtat the index of the pokemon you want to delete is, once you have lots of todos this will get very hard

Make sure to think about how the users will use the app and try to design for this :)


But in general it's super cool that you added this!! Love the little pokemon drawings :))))

File: src/ex3/cli-methods.js [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+    if(!indexValidation(getIndexData)) return
+
+    const data = main.itemManager.getTodoList()
+    const {index} = getIndexData
+    
+    if(!data[index].isPokemon) {
+      console.log(chalk.bold.red("that's is not a Pokemon"))
+      return
+    }
+
```
Comment (shaneer) [inline src/ex3/cli-methods.js:302]: would have been better to put this into a constant and then havea function that builds the image with this const

File: src/ex3/item-manager.js [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+    addTodo(enterValue){
+        const trimValue = ItemManager.trim(enterValue)
+
+        this.handleInputToAdd(trimValue)
+    }
+
+    handleInputToAdd(trimValue){
+        const partOfNumbersSeprateWithComma = 
+            trimValue.substring(0,1).match(multiNumbersSeparatedWithComma)
```
Comment (shaneer) [inline src/ex3/item-manager.js:28]: this just checks the first char, what if the rest is not a number? This doesn't confirm that the string is numbers separted by commas
Also you can jsut use isNumber if you're just checkihng one char this is enough

File: src/ex3/item-manager.js [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+
+        const singleNumberPattern = trimValue.match(singleNumber)
+        const singleWordPattern = trimValue.match(singleWord)
+        const multiNumberPattern = trimValue.match(multiNumbersSeparatedWithComma)
```
Comment (shaneer) [inline src/ex3/item-manager.js:32]: you can make a regex that would match multiple cases instead of doing 4 separate checks, also it doesn't seem like you do anything with these checks later separately rather if any of them is true then it's the same function so this would make more sense

File: src/ex3/item-manager.js [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+        }
+        else{ //noraml todo
+            const isPokemon = false
+            const imagePokemonPath = null
+            this.addTodoParse(trimValue, isPokemon, imagePokemonPath)
+            this.updateTodos()
+        }
+    }
+
+    handleAddSingleOrMultiPokemonsTodo(enterValue){
```
Comment (shaneer) [inline src/ex3/item-manager.js]: separate this to a file that could be pokemonUtils or handlers or something like that

File: src/ex3/item-manager.js [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+    updateTodos(){
+        this.model.saveDataToLS()
+    }
+
+    getDataInIndex(index){
+        if(index < 0 || index >= this.model.todoList.length){
+            return null
+        }
+
+        return this.model.getDataInIndex(index)
```
Comment (shaneer) [inline src/ex3/item-manager.js:78]: you could validate the index here and then return null if it doesn't exist and just check that outside

File: src/ex3/cli-methods.js [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+      name: 'editTodoIndex',
+      message: 'enter the index for todo that you want to edit', 
+    })
+    .then(answers => {
+      if(!numberValidation(answers.editTodoIndex)) return
+      
+      const getIndexData = main.getDataInIndex(answers.editTodoIndex)
+      
+      if(!indexValidation(getIndexData)) {
+        return
```
Comment (shaneer) [inline src/ex3/cli-methods.js:140]: would be better to communicate this

**Comment by shaneer** [comment]
In general the solution worked well and it was cool to see all the extras you added :)
When reviewing the code it was hard to follow the flow and give a full code review as the code is a little jumbled at the moment, lots of really long files that have many different unrelated functions and it seems you still don't have prettier or some other autoformatter installed which leaves a lot of diffs in punctuation and spacing that make the code a little less readable.

I would really suggest configuring something that autoformats on save as well as separating your code into modules/services by responsibility and trying to keep files smaller that way the code review can be more efficient as well as it will be easier to look at the flow and focus on logical issues

**Comment by Ouriel91** [comment]
Thanks about all comments @shaneer 
i'll fix things by your comments, there are some things that i'll take the improvments from ex3 toex4 (unfortonally most of them not because the ex3 and ex4 are totally diffrent excercises)

thanks it was really helpful
